Fixes #7119 - Add support for Facebook Flow language

jdelStrother opened the pull request:
--------------------------------
**👉  There is a bounty source opened on this issue, click on the badge for more info**

[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=37889526)](https://www.bountysource.com/issues/37889526-some-request-support-facebook-flow-language?utm_source=37889526&utm_medium=shield&utm_campaign=ISSUE_BADGE)

------------------------------

Feature list in order to complete this PR and unlcok the bounty:

**Auto-completion**
- [ ] The layer should activate `company-flow` whenever the buffer begins with either of these lines:
    - `// @flow`
    - `/* @flow */`

[Author note: It's currently enabled in all of react-mode and js2-mode, and I've not noticed any ill-effects in files where flow isn't present.]

**Syntax Highlighting**
- [ ] The layer should improve syntax highlighting for `flow`. Right now it's mostly ok in `react-mode` and quite bad in `javascript-mode` (for instance, use of generics `Map<K, V>` breaks the syntax highlighting). There may be some inspiration in `emacs-flow-jsx`, `react-mode` and `typescript-mode`.

**Documentation**
- [x] The type of current word should be rendered with `eldoc` -  `flow type-at-pos` should give the type.
- [x] bonus points for highlighted `eldoc` like in `typescript`.

**Code Navigation**
- [x] Using `flow get-def` and `flow find-module` should enable amazing jump to definition functionality. I'd love to have them binded to <kbd>g</kbd> or <kbd>d</kbd> or at least local leader.

-----------------------------

*Original message*

Hi there - I've cobbled this together from the various flow-related things I've been using. There seems to be some interest in having a proper spacemacs flow layer (#7119), so thought I'd throw this out for discussion.

There's a few bits that might be somewhat targeted towards my own setup - for example, I chain the flow flychecker onto the eslint checker, but conceivably others might want to chain it onto jshint instead.   How much of that would you want to see customizable before merging this?

-----------------------------

EDIT: [@syl20bnr] changed title and message body to reflect the bounty source attached to it.

pheuter commented:
this is great, thanks so much for creating a layer, I use flow every day and now I can switch back to Emacs for my web development.

Fwiw, I was only able to get the proper `company-flow` completions to show up when I disabled `tern-mode`:

```elisp
(remove-hook 'react-mode-hook 'tern-mode)
```

Otherwise, it was only populating completions from tern instead of flow.

jdelStrother commented:
@pheuter ah, I actually have tern disabled globally via `dotspacemacs-excluded-packages`.

If people are using tern, do you think they'd prefer to have flow completions or tern completions?   I'm not sure how good tern completions usually are.   (I'm also not sure how I'd disable tern based on the presence of the `@flow` directive, so if anyone has any suggestions...)

jdelStrother commented:
Discovered that `eldoc-documentation-function` would potentially hang emacs for several seconds if you made a change to a file that required flow to restart its server.  I've changed the flow invocation there to `flow --retry-if-init=false` and rebased onto the latest develop.

evindor commented:
I've started a bounty for this feature in #7119 [![Bountysource](https://api.bountysource.com/badge/issue?issue_id=37889526)](https://www.bountysource.com/issues/37889526-some-request-support-facebook-flow-language?utm_source=37889526&utm_medium=shield&utm_campaign=ISSUE_BADGE)

@aaronjensen pointed out that the "whole thing has suffered from feature creep, and it's unclear what the bounty is for".

The bounty is for improving, and then merging, this pull request. Which effectively closes #7119.

I suggest the following list to consider the bounty requirements satisfied:
- (autocompletion) The layer should activate company-flow company-mode whenever the buffer begins with `// @flow` or `/* @flow */`
- (syntax) The layer should improve syntax highlight for flow. Right now it's mostly ok in react-mode and quite bad in javascript-mode. (for me use of generics `Map<K, V>` breaks the syntax after this line) There may be some inspiration in [emacs-flow-jsx](https://github.com/omouse/emacs-flow-jsx), react-mode and typescript-mode.
- (eldoc) The type of current word should be rendered with eldoc, bonus points for highlighted eldoc like in typescript. `flow type-at-pos` should give the type
- (jump-to-definition) Using `flow get-def` and `flow find-module` should enable amazing jump to definition functionality. I'd love to have them binded to `g d` or at least local leader.

With these it would become very comfortable working with flow in spacemacs. There are a bunch of interesting extras like `flow find-refs` and `flow coverage`.

aaronjensen commented:
> (autocompletion) The layer should activate company-flow company-mode whenever the buffer begins with // @flow or /* @flow */

This would require work in [`company-flow`](https://github.com/aaronjensen/company-flow) because right now it is enabled on a major mode basis. If there was a major mode for flow, then we could just use that, but it would need to be derived from something (I'd prefer `rjsx-mode`, but others may prefer `js2-mode`), so it's complicated. Maybe `flow-mode` should be a minor mode? I wouldn't be sure how to only enable that in buffers w/ the `@flow` attr. Also, basing it on `@flow` will break for people that have `[options] all=true` in their `.flowconfig`. I'd probably prefer to just look for `.flowconfig`.

jdelStrother commented:
@evindor do you have a fuller example of highlighting not working with generics?   They seem ok to me in js2-mode.

evindor commented:
@jdelStrother  I've run a few experiments and looks like reduced the problem to a few edge cases:

### Argument type annotations
No Flow - all good:
<img width="277" alt="screen shot 2017-01-29 at 21 05 50" src="https://cloud.githubusercontent.com/assets/822951/22406612/eaae4974-e666-11e6-8f1e-832c076bf263.png">

Add some typing:
<img width="365" alt="screen shot 2017-01-29 at 21 06 37" src="https://cloud.githubusercontent.com/assets/822951/22406618/16375a54-e667-11e6-8741-454c330a871d.png">
Notice how arguments to both functions lose the highlight

### Generic in the argument type annotation:
All good:
<img width="468" alt="screen shot 2017-01-29 at 21 10 56" src="https://cloud.githubusercontent.com/assets/822951/22406641/7bc3579c-e667-11e6-83c7-06b59d21c20f.png">

Let's break it:
<img width="442" alt="screen shot 2017-01-29 at 21 12 42" src="https://cloud.githubusercontent.com/assets/822951/22406650/bb7374ee-e667-11e6-9eca-398bf347d42b.png">
@jdelStrother look how almost all highlight after `Handler<T>` declaration is gone.
```javascript
export type Handler<T> = {
  [key: string]: (state: T, action: Action<any>) => T,
}                                       // ^^^ This is the troublemaker
```

Fixable by extracting this `Action<any>` to the outside:
<img width="425" alt="screen shot 2017-01-29 at 21 16 51" src="https://cloud.githubusercontent.com/assets/822951/22406691/72e20f32-e668-11e6-8f86-a8f6860552f3.png">

evindor commented:
I don't think these are the only cases, I'll get back to work and if I see anything else I'll report here. Also, in general, can we have some kind of highlight for the annotations themselves?

Here I rewrote this sample piece to **typescript**:
<img width="486" alt="screen shot 2017-01-29 at 21 29 28" src="https://cloud.githubusercontent.com/assets/822951/22406768/16827702-e66a-11e6-8355-68a523faa113.png">

evindor commented:
Regarding eldoc, check out this **typescript** glory:

<img width="261" alt="screen shot 2017-01-29 at 21 34 31" src="https://cloud.githubusercontent.com/assets/822951/22406825/fc5aac40-e66a-11e6-9976-39efe6ad7117.png"><img width="256" alt="screen shot 2017-01-29 at 21 35 10" src="https://cloud.githubusercontent.com/assets/822951/22406826/fe7833d0-e66a-11e6-9226-f6bf93198ddd.png">

jdelStrother commented:
Yeah, I looked at the code the typescript package uses to produce that.
It's... not trivial
On Sun, 29 Jan 2017 at 18:37, Arseny Zarechnev <notifications@github.com>
wrote:

> Regarding eldoc, check out this *typescript* glory:
>
> [image: screen shot 2017-01-29 at 21 34 31]
> <https://cloud.githubusercontent.com/assets/822951/22406825/fc5aac40-e66a-11e6-9976-39efe6ad7117.png>[image:
> screen shot 2017-01-29 at 21 35 10]
> <https://cloud.githubusercontent.com/assets/822951/22406826/fe7833d0-e66a-11e6-9226-f6bf93198ddd.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/syl20bnr/spacemacs/pull/7208#issuecomment-275935427>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAJSXHF_KpWt8-ly2n3jyI9tbhVeaUyks5rXNxRgaJpZM4KIoh5>
> .
>

evindor commented:
@jdelStrother well, these are bells and whistles, can live w/o fancy colored eldoc.

evindor commented:
@jdelStrother for the same piece of code, tried `react-mode`:
<img width="437" alt="screen shot 2017-01-29 at 21 50 09" src="https://cloud.githubusercontent.com/assets/822951/22406908/0a335522-e66d-11e6-8479-cf1352021bcd.png">

Look like the highlight **isn't screwed up anywhere!** I've looked into [react-mode](https://github.com/syl20bnr/spacemacs/blob/master/layers/%2Bframeworks/react/packages.el#L78) and they don't define any extra syntax, only derive from `web-mode`.

Then I switched to `web-mode` and the code was still looking like on the screenshot above (taken in `react-mode`)

evindor commented:
Please excuse me for linking here cause it's not very readable, but I guess there should be some inspiration available [here in web-mode](https://github.com/fxbois/web-mode/blob/master/web-mode.el#L1652)

jdelStrother commented:
I think fixing up js2-mode's highlighting for flow is probably beyond me - their parser is pretty terrifying.
However, I did manage to get highlighting working in eldoc using web-mode's highlighter - will push a fix for that later.

evindor commented:
Hey @jdelStrother I'm trying this out (just checking out my `.emacs.d` to your branch) and spacemacs becomes really slow on javascript files, plus it looks like flycheck checkrs never complete and show `?` in the status bar. Am I doing something wrong?

jdelStrother commented:
One possibility that jumps to mind is that maybe you have competing flow versions in emacs and your terminal, which results in the flow server restarting itself all the time.

Try this, just for debugging's sake:

- Open a javascript file in emacs
- Switch to your terminal and run `flow stop` to ensure there's no flow processes lying around
- Switch to emacs and hit `,fs` (`flow-type/show-start-server`) to launch flow in a window in your emacs session.

See if that makes any errors more obvious?

evindor commented:
I think the movement became slower as well, maybe because calls to `flow-type/type-at-cursor` are not throttled?

evindor commented:
I've followed your advice, made sure the server is started from within emacs. Then also recorded a gist, every 2 seconds [`ps aux | grep flow`](https://gist.github.com/evindor/83cc8e280d980b59b395c0fe4fc5a326)

evindor commented:
While recording that I just jumped around various files, waited for eldoc to show up or pressed `g d` a couple times. Also here is the log from emacs [log buffer](https://gist.github.com/evindor/bcb7c96511e6ea3d790d31ad433ea8d8).

evindor commented:
Sorry that I only mentioned the performance issues, @jdelStrother otherwise it's starting to look great! `g d` jumps, eldoc is glorious and completion seems to hook.

evindor commented:
Does it respect `.flowconfig` in the root of my repo? It feels like every time I open a file the whole process starts anew - and I have to wait for flow to `check-contents` to get completions and types and everything. Also, whenever I manage to kill all flow servers and start one with `, f s`, it eventually dies:
```
[2017-02-07 00:29:05] Initializing Server (This might take some time)
[2017-02-07 00:29:05] executable=/Users/evindor/Sites/some/SomeApp/node_modules/flow-bin/flow-osx-v0.37.4/flow
[2017-02-07 00:29:05] version=0.37.4
[2017-02-07 00:29:05] Parsing
[2017-02-07 00:29:12] Building package heap
[2017-02-07 00:29:13] Loading libraries
[2017-02-07 00:29:14] Running local inference
[2017-02-07 00:29:19] Re-resolving directly dependent files
[2017-02-07 00:29:19] Calculating dependencies
[2017-02-07 00:29:19] Merging
[2017-02-07 00:29:28] Done
[2017-02-07 00:29:28] Server is READY
[2017-02-07 00:29:28] Took 22.998838 seconds to initialize.
Status: Error
flow server is out of date. Exiting.

Process flow ~/Sites/some/SomeApp/ exited abnormally with code 9
```

aaronjensen commented:
@evindor it looks like you have a configuration issue. Likely you have two versions of flow floating around (maybe one installed globally and one in your node_modules). I'm not sure what would be switching between the two, but something probably is. I believe this is what I see when my global flow version does not match my version in the project.

jdelStrother commented:
Do you have anything else that might be launching flow? I've seen that sort
of behaviour when emacs keeps launching it via node_modules/.bin/flow in
your project directory, but something else might launch it via a globally
installed flow, and the two versions conflict and cause flow to restart the
server each time.
On Mon, 6 Feb 2017 at 21:37, Arseny Zarechnev <notifications@github.com>
wrote:

> Does it respect .flowconfig in the root of my repo? It feels like every
> time I open a file the whole process starts anew - and I have to wait for
> flow to check-contents to get completions and types and everything. Also,
> whenever I manage to kill all flow servers and start one with , f s, it
> eventually dies:
>
> [2017-02-07 00:29:05] Initializing Server (This might take some time)
> [2017-02-07 00:29:05] executable=/Users/evindor/Sites/some/SomeApp/node_modules/flow-bin/flow-osx-v0.37.4/flow
> [2017-02-07 00:29:05] version=0.37.4
> [2017-02-07 00:29:05] Parsing
> [2017-02-07 00:29:12] Building package heap
> [2017-02-07 00:29:13] Loading libraries
> [2017-02-07 00:29:14] Running local inference
> [2017-02-07 00:29:19] Re-resolving directly dependent files
> [2017-02-07 00:29:19] Calculating dependencies
> [2017-02-07 00:29:19] Merging
> [2017-02-07 00:29:28] Done
> [2017-02-07 00:29:28] Server is READY
> [2017-02-07 00:29:28] Took 22.998838 seconds to initialize.
> Status: Error
> flow server is out of date. Exiting.
>
> Process flow ~/Sites/some/SomeApp/ exited abnormally with code 9
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/syl20bnr/spacemacs/pull/7208#issuecomment-277821583>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAJSd7LP17YHdZm0ojwcKFv8wIedHr3ks5rZ5J8gaJpZM4KIoh5>
> .
>

evindor commented:
@jdelStrother @aaronjensen thanks guys, I'll `brew unlink flow` right now and check it out

jdelStrother commented:
@aaronjensen got any bright ideas on handling that case better?  I'm guessing it's a pretty common setup, though maybe less and less now that the official recommendations are to install via `npm install flow-bin`

aaronjensen commented:
@jdelStrother For one, this layer should ensure that all pieces of it are using the same flow exec. If some use global and others use node_modules due to paths being set up differently or something it'd exacerbate the problem. I don't know what else would cause it. I saw it when I ran `$ flow` from my terminal if it was a different version, that'd cause emacs to start the server over next time.

evindor commented:
So even without a global flow binary, something is starting flow on emacs startup. I'm seeing the home buffer, no files opened yet and yet `ps aux` shows a local flow started.

evindor commented:
For some reason tern dominates my autocomplete. Unless I explicitly turn it off, I either get a wild mix of tern and flow completion or nothing at all.

Flow-enabled js file, `company-diag` output:
```
Emacs 25.1.1 (x86_64-apple-darwin16.0.0) of 2016-09-30 on mbp13.local
Company 0.9.2

company-backends: (company-tern company-flow
              (company-dabbrev-code company-gtags company-etags company-keywords)
              company-files company-dabbrev)

Used backend: company-tern
Prefix: ("" . t)

Completions:
  #("0" 0 1 (isProperty t name #0 type "module.exports|[?]" depth 0)) " -> module.exports|[?] ○"
  #("1" 0 1 (isProperty t name #0 type "[module.exports|[?]]" depth 0)) " -> [module.exports|[?]] ○"
  #("@@functional/placeholder" 0 1 (isProperty t name #0 type "bool" depth 0)) " -> bool ○"
.......
```

aaronjensen commented:
@evindor a few questions...

> So even without a global flow binary, something is starting flow on emacs startup. I'm seeing the home buffer, no files opened yet and yet ps aux shows a local flow started.

1. What does this mean to you? Is it somehow related to your previous problems?
2. Did you disable resume layouts on spacemacs, quit emacs entirely, stop/kill flow and ensure it wasn't running before restarting spacemacs?

aaronjensen commented:
> For some reason tern dominates my autocomplete.

Are you copying issues from https://github.com/syl20bnr/spacemacs/issues/7119 ? You've mentioned this there and it has been discussed. I believe it is not yet fixed.

jedahu commented:
eldoc slows things down for me. I've disabled it for flow and instead use this command bound to ",ft".

```elisp
(defun flow-type/show-type-at-cursor ()
  (interactive)
  (message "%s" (flow-type/type-at-cursor)))
```

flycheck-flow's `javascript-flow` checker will not start flow if `flycheck-javascript-flow-args` contains `--no-auto-start`. I've just made a pull request for `javascript-flow-coverage` to use this arg too.

The flow-type layer adds `--no-auto-start` to `flycheck-javascript-flow-args` if `flow-type/no-auto-start` is not nil. This only happens on layer init though.

jdelStrother commented:
If, like jehuda, you're finding the type-at-cursor stuff too slow, you can disable it with this in your `dotspacemacs-configuration-layers` : 

```emacs
(flow-type :variables flow-type-enable-eldoc-type-info nil)
```

tko commented on layers/+lang/flow-type/funcs.el:
```diff
@@ -0,0 +1,133 @@
+(defun flow-type/flow-binary ()
+    (or
+     ;; Try to find flow in node_modules (via 'npm install flow-bin')
+      (let ((root (locate-dominating-file buffer-file-name "node_modules")))
```
Instead of duplicating this logic in multiple(?) layers maybe consider `add-node-modules-path` as in  https://github.com/syl20bnr/spacemacs/pull/8621

jdelStrother replied:
Sounds good, will give it a try

jdelStrother replied:
I *think* the best way of using this (once #8621 is merged) is to assume that `js2-mode` / `react-mode` have already included `add-node-modules-path` and called it in their setup hooks.   Then we can just delete the `Try to find flow in node_modules` block of code from here.

jdelStrother commented on layers/+lang/flow-type/README.org:
```diff
@@ -0,0 +1,38 @@
+ - Add flow checking to flycheck (based on the flycheck-flow package)
+ - company-mode completion using company-flow
+ - "compilation" using =flow status= and =flow check=
+ - auto staring =flow server= in a project specific comint buffer
```
Agreed

jdelStrother commented on layers/+lang/flow-type/README.org:
```diff
@@ -0,0 +1,38 @@
+To start the server in a comint buffer on entering a =js2-mode= buffer, set it to ='process=.
+
+#+BEGIN_SRC emacs-lisp
+(flow-type :variables flow-type-no-auto-start 'process)
```
Urgh, possibly.   I have two counter arguments:


1) `no-auto-start` matches the `flow` option  - from `flow --help`: 

```
  --no-auto-start  If the server is not running, do not start it; just exit
```

2) In its current form, the two possible truthy values for `flow-type-no-auto-start` ( `'process` & `t`)  are siblings - they both add `--no-auto-start` to the flow command.  With your suggestion, it'd be `'process` & `nil` that add `--no-auto-start` - one truthy and one falsey value, which feels a little weird.


They're not strong arguments though - if you still think it'd be better the other way round, I'll change it.

aaronjensen replied:
What is the reason for wanting to set this in the first place?

jdelStrother replied:
/cc @jedahu ?

tko replied:
I think it's better to phrase it as positive for presumably better ease of use, but I can totally understand the ease of implementation argument, gone that way many times myself. Just a random suggestion, shouldn't be a showstopper either way.

aaronjensen commented:
Now that https://github.com/an-sh/flow-minor-mode is a thing, should this use that instead of defining its own functions for some of the common operations?

jdelStrother commented:
@aaronjensen almost certainly.   I'll try and upstream a few of my changes into that.

jdelStrother commented:
I've started a fresh branch at https://github.com/jdelStrother/spacemacs/tree/flow-type-minor which is re-implemented on top of flow-minor-mode

jedahu commented:
@aaronjensen, you might be running flow server manually in a terminal for
some reason.


On 15 Apr 2017 02:35, "Aaron Jensen" <notifications@github.com> wrote:

*@aaronjensen* commented on this pull request.
------------------------------

In layers/+lang/flow-type/README.org
<https://github.com/syl20bnr/spacemacs/pull/7208#discussion_r111581268>:

> + - Show the flow-type of the object under the cursor using Eldoc
+ - Add flow checking to flycheck (based on the flycheck-flow package)
+ - company-mode completion using company-flow
+ - "compilation" using =flow status= and =flow check=
+ - auto staring =flow server= in a project specific comint buffer
+
+* Install
+To use this configuration layer, add it to your =~/.spacemacs=. You
will need to
+add =flow-type= to the existing =dotspacemacs-configuration-layers=
list in this
+file. Please ensure you already have the flow binary installed and in
your PATH.
+
+To prevent the flow server from auto-starting, set
=flow-type-no-auto-start= to =t=.
+To start the server in a comint buffer on entering a =js2-mode=
buffer, set it to ='process=.
+
+#+BEGIN_SRC emacs-lisp
+(flow-type :variables flow-type-no-auto-start 'process)

What is the reason for wanting to set this in the first place?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/syl20bnr/spacemacs/pull/7208#discussion_r111581268>,
or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAGJCwDYmFhuoXyNgbkvXM6wwrsXBnpfks5rv4QkgaJpZM4KIoh5>
.

aaronjensen commented:
> @aaronjensen, you might be running flow server manually in a terminal for
> some reason.

Not sure I follow. If you're running it manually in a terminal it will detect that and use it. The only reason I can think of to not auto start is if you'd rather have it fail than take on the time to start the server. Am I missing something?

jedahu commented:
@aaronjensen I've had trouble before where the terminal environment was not the same as the emacs one, so when emacs called flow it didn't find the existing server and started fresh one.

z3t0 commented:
How do I add this to my spacemacs setup?

orther commented:
@z3t0 I personally just replaced my 2 year spacemacs confit with @aaronjensen 's setup and I'm loving life. Yeah a bit lazy and some of my custom stuff got bumped but for now it's a great way to go. Besides that there are several examples above. Would be happy to help.

orther commented:
PS as a bonus w/ prettier auto formatting. That pretty much life changing and integrate with flow.

z3t0 commented:
@orther I'm not seeing the setup, could you share a link? thanks

orther commented:
This is his repo I'm referring to. It also fixed my Elixir formatting. Not to meantion the rjsx mode for react and JS is better imo

orther commented:
https://github.com/aaronjensen/spacemacs.d/pulls Sorry forgot link

z3t0 commented:
got it thanks

skattyadz commented:
Just set up [flow-type-minor](https://github.com/jdelStrother/spacemacs/tree/flow-type-minor) is and it's working well with [aaronjensen's rjsx mode](https://github.com/aaronjensen/spacemacs.d/tree/master/layers/%2Baj/aj-javascript). 

I do get a small lag for about a second just after opening a file though. I can live with it for now though. My config's at https://github.com/skattyadz/dotfiles/tree/master/.spacemacs.d/layers

z3t0 commented:
@skattyadz Thank you for the link but I am getting the following [error](https://i.imgur.com/UySv0Ki.png)

Sorry it's an image file because I couldn't figure out how to keep the error there long enough to copy it.

DrPandemic commented:
I would like to add my grain of salt here. I have been playing with this PR for maybe one week. Before that I was using [my mode](https://github.com/DrPandemic/dotfiles/tree/master/spacemacs.d/private/drpandemic-javascript) which is mostly copy pasted code from you and aaronjensen. Your mode is pretty good for my use cases, but I would like to point out some of my pain points. Note: most of my testing was made with `react-mode` and all my issues are related to `react-mode`.

## Company
By default with `js2-mode`, `company-flow` is present in `company-backends`. But when I switch to `react-mode` the flow backend is not there anymore. Earlier in this PR, you talked with @pheuter about `tern-mode` vs `company-flow`. IMHO, if flow is working `company-flow` is highly superior to `tern-mode`. It gives better situational content, less useless completion and documentation.

## Major
I played with `js2-mode`, `rjsx-mode` and `react-mode` and I think that `react-mode` has the best syntax coloring and is the most stable. I have some issues with `react-mode`, but less than with the others. So my question is, should this mode set a default major mode for flow/javascript files? If it does, I would argue that `react-mode` should be chosen.

## Eldoc
Eldoc is not present with `reac-mode`.

Fuco1 commented:
FWIW I've started some work (we're tracking it https://github.com/antifuchs/rjsx-mode/pull/1) to extend `js2-mode` parsing to support flow... so this will provide you with complete AST as well as all the goodies of `js2-mode`.  It's mostly usable for the 90% use-case now.  In the coming days I will turn it into a separate package as there is no relation to `rjsx` (react).

It will provide syntax highlighting as well just like `js2-mode` does.

syl20bnr commented:
Any news about this bounty ?

@Fuco1 created a repository there: https://github.com/Fuco1/flow-js2-mode

Fuco1 commented:
I don't know how to integrate it with spacemacs nor will I try to do that anytime soon but if anyone wants to use the package I've created feel free to do so.

smile13241324 commented:
@jdelStrother I have merged the first draft version of your layer provided #10513. I think this should be working as a first draft however to get full ide features I would recommend adding a proper LSP server for this.

Anyway this merge made this PR obsolete.

jdelStrother commented:
> @jdelStrother I have merged the first draft version of your layer provided #10513. I think this should be working as a first draft however to get full ide features I would recommend adding a proper LSP server for this.

Yeah, I've been experimenting with that recently - possibly we should just delete this layer entirely and suggest that people use:

```emacs-lisp
dotspacemacs-configuration-layers '(
  (javascript :variables
      javascript-backend 'lsp))
```

?

wukkuan commented:
@jdelStrother - I think that sounds like a much better solution, FWIW.

smile13241324 commented:
@jdelStrother @wukkuan ok so this is included in the current lsp implementation, I haven't been sure about this, then it would suffice to add some hints to the javascript layer's documentation and let users use lsp if they require flow support. Care to make a PR?

wukkuan commented:
@smile13241324 - I'm willing to help out with documentation at the very least. But I'm not able to find any way to enable LSP currently. Am I missing something?

smile13241324 commented:
@wukkuan - thanks for doing the PR :), and sorry for my late answer I am currently pretty busy at work so there is not as much time for spacemacs as I would like.
